Port/1.20.4

Nightenom opened the pull request:
# List of missing things:
- runtime testing :)
  - [ ] NEED: test that upgrading/repairing building doesn't break rot/mirror, same for loading/unloading/world upgrading
  - [ ] Yes I tested this before submitting it.
  - [ ] I also did a multiplayer test.
- [x] datagen - ideally should create smallest diff as much as possible (known breaks: DO blocks) so we dont break datapacks, review https://github.com/ldtteam/minecolonies/commit/d15bb334f60b62f0db754c3b8110778dd7ed39d1 doesnt break things
- [x] some textures need unstitching or moving to atlas location (with unstitch mapping) commit: https://github.com/ldtteam/minecolonies/commit/5c99cf8a60bb4154a4829b310e83b4987eaa1662
- sanity checks of following commits: 
  - [ ] https://github.com/ldtteam/minecolonies/commit/5cd5afc27c043279086c72422fa3f46f4871a203 - does this break world data (= how often does the attributes reapply)?
  - [ ] https://github.com/ldtteam/minecolonies/commit/91026a03fd2056bae6db9cd7f6c54973b31d3dac
  - [ ] [replacement for icapabilityprovider since we use itemhandler cap only anyway](https://github.com/ldtteam/minecolonies/blob/port/1.20.4/src/main/java/com/minecolonies/api/util/IItemHandlerCapProvider.java)
  - [x] (self checked) https://github.com/ldtteam/minecolonies/commit/e67e891fd4daf878185f68da25f26bc9082803ab block codecs aren't used now but we should prepare for another porting pita
  - [x] https://github.com/ldtteam/minecolonies/commit/89c7b0d2c9a16b156bbaa9e938361d9030ae6a07 some weird ways to guess rotation eventhough we have it? (dont fully understand what is goal of this code)
  - [ ] https://github.com/ldtteam/minecolonies/commit/c9451c06edc65fbd10d4e73ee8a646c64d5ffbab entity spawn packets not being needed for entities that dont do custom data? 
- #9771

Review please

Nightenom committed:
class remapper

Nightenom committed:
gradle 8.6

Nightenom committed:
versions, cleanup gradle code

Nightenom committed:
config, LanguageHandler -> blockui.common

Nightenom committed:
registry

Nightenom committed:
mod ctor

Nightenom committed:
AT - first pass

Nightenom committed:
structurize imports

Nightenom committed:
Pay respects to PlacementSettings, Rotation and Mirror

Nightenom committed:
Networking: messages

Nightenom committed:
Network: call sites

Nightenom committed:
some registry/network leftovers

Nightenom committed:
actual porting, few leftovers from prev commits, AT sec pass

Nightenom committed:
deprecations

Nightenom committed:
Apply #9660

Nightenom committed:
Recipes

Nightenom committed:
advancements (includes refactor so flatten post-porting folders with one file)

Nightenom committed:
Actual porting

Nightenom committed:
Review: attribute serialization changed - do we care?

Nightenom committed:
Sanity review check: backuphelper to nio Path

Nightenom committed:
Needs fix in assets: prepare for unstitching - will crash now at runtime so we dont forge to fix

Nightenom committed:
Update EmptyColonyStructure.java

Nightenom committed:
AT cleanup

Nightenom committed:
Hopefully correct atlas for current atlas textures

Nightenom committed:
xml fixes

Nightenom committed:
Amend 937e33f - miner nodes rotation mirror

Raycoms committed:
fix 9774 (#9775)

Better handle removed quests
Fix trapdoor handling with ladders
Fix some panel/carpet movement.

Nightenom committed:
ItemHandler caps

Nightenom committed:
Block codecs

Nightenom committed:
Actual porting

Nightenom committed:
Amend 937e33f - buildings having their super weird rotation hackery

Nightenom committed:
PlayerInteractEvent issues, please rework this

Nightenom committed:
Few potentially insecure loottables porting - we'll see during datagen

Nightenom committed:
Entity networking

Nightenom committed:
Actual porting - last things

Nightenom committed:
journeymap

Nightenom committed:
move caps to single package

Nightenom committed:
port capabilities (without client sync for now)

Nightenom committed:
fix datagen

Nightenom committed:
Runtime fixes

Nightenom committed:
datagen

Nightenom committed:
Various data/assets fixes

Nightenom committed:
Runtime bugs sec edition

Thodor12 committed:
Update jpegs to pngs (#9783)

Update jpegs to pngs

MotionlessTrain committed:
Urban Savannah and Colonial fixes in 1.20.1 (#9777)

Fix broken DO blocks in the Urban Savannah Hillside Residence
Fix some glitched racks in the Colonial level 2 warehouse

Nightenom committed:
colony + colonyManager to ServerLevel

Update jpegs to pngs

Nightenom committed:
Runtime fixes 3

Nightenom committed:
fix npe

Nightenom committed:
Fix caps

Nightenom committed:
simplify rotmir logic

Nightenom committed:
colors in xml, nulls in interactions

Raycoms committed:
push changes

someaddons committed:
Fix double feeding confusion (#9784)

Add statistics to herders
Merge feeding children and normal feeding

MotionlessTrain committed:
Supply camp and ship enhancement (#9779)

Made survival friendly versions of supply camps and ships

Raycoms committed:
fix wayyyyy too large pngs

Raycoms committed:
Some fixes (#9796)

Make sure to reset the position after the water step.
Properly handle enchantment damage etc
Don't break blocks when turned off when placing ladders
Don't get stuck in supplies window when selecting a wrong pack once.

someaddons committed:
Add set rank command (#9789)

Add a set rank command

Thodor12 committed:
Fix armor level checking (#9786)

When checking for armor level, the entire armor set + toughness is taken into consideration as opposed to only the armor level of the chestplate.
Fixed the MinecoloniesArmorMaterial to updated ArmorMaterials Mojang implementation, so that armor values are applied in the correct order to each armor type.
Updated plate armor:
Durability raised from 20 to 37
Armor levels updated from 4,9,7,4 to 3,8,6,3 (helmet, chestplate, leggings, boots)
Toughness reduced from 2 to 0
Enchantment level raised from 6 to 9
Updated pirates armors:
Pirate armor 1 is now identical to iron with durability of diamond
Pirate armor 2 is now identical to diamond with durability of iron
Fixed other armors in the mod to use the new armor material constructor (santa hat and build goggles)

MotionlessTrain committed:
Medieval Oak and Minecolonies Original fixes (#9797)

Medieval Oak:
Fixes the moving rack in the builder's hut from level 4 to level 5 (the level 4 rack is now a block higher, matching the position at level 5)
Minecolonies Original:
Add the missing furnace at the level 1 dyer's hut

someaddons committed:
pathfinding adjustments (#9798)

Add more distance utils to make them easier to use
Fixed a couple bugs where wrong positions were passed
Switch pos to ints within pathfinding to lower memory overhead
Add shape utils for faster shape checking

MotionlessTrain committed:
Fix the Stalactite Caves barracks having an end stone block at levels 3 and 4 (#9808)

Fix the barracks having an end stone block at level 3 and 4

Raycoms committed:
fix server loading

Raycoms committed:
fix message and sidedness issues

Raycoms committed:
fix messages

Raycoms committed:
more fixes

Raycoms committed:
fix

someaddons committed:
Fix visitor sitting position being off at large coordinates (#9813)

Fix visitor sitting position being off at large coordinates

Raycoms committed:
several fixes

Raycoms committed:
fix whitespace

imgbot[bot] committed:
[ImgBot] Optimize images (#9799)

Your image file size has been reduced by 6%

Raycoms committed:
some fixes

Raycoms committed:
fix 9822

Raycoms committed:
less nulls

Thodor12 committed:
Fix second crash in townhall map GUI (#9823)

Remove unneeded call to removeChild

MotionlessTrain committed:
Minecolonies Original and Stalactite Caves fix (#9818)

Fix the staircase of the Minecolonies Original residence level 5
Fix the racks in the Stalactite Caves builder's hut

Raycoms committed:
fix

Thodor12 committed:
Fix stonemason not being able to craft decorated pots, remove dyeing recipes from stonemason (and other buildings where same issue may exist) (#9827)

Ensure that the stonemason can craft using brick (not just bricks, the full block)
Added dyer recipes to product inclusions, so that these are not references by other buildings (prevents i.e. stonemason from dyeing bricks)

Raycoms committed:
fix field message

Raycoms committed:
translatable escape

Thodor12 committed:
Update issue templates (#9838)

Renames alpha to beta

Raycoms committed:
bunch of fixes (#9832)

Not so strict on guards not eating/sick handling (as we already got an override on the AI anyway)
Make sure builders recalc on restart
Try to improve compat for enchantments, etc
Fix nametag rename
Fix crash in research window

Talyda committed:
Update housing2.json (#9834)

Update housing2.json

someaddons committed:
fix path reaching value (#9840)

Fix the finalized path not getting the correct reached flag, as the result returns it only when its processed

Raycoms committed:
fix

someaddons committed:
Hotfix lumberjack pathfinding

MotionlessTrain committed:
Shire fix and Medieval adjustments (#9846)

Fix the water feature in the Shire flower shop
Fix relative offsets of university and library in the MedievalOak universitylibrary decoration (the wrong pillar was overlapping, which gave the library an offset of five blocks)
Make sure the medieval universitylibrary decorations have level 0 hut blocks in them

Raycoms committed:
try to fix this

Raycoms committed:
rollback

Raycoms committed:
fix

Raycoms committed:
fix to map sync (#9853)

More crash tolerant map synching

uecasm committed:
Fix infinite debugging (#9856)

Fix infinitely long names when state debugging is enabled

Raycoms committed:
Merge branch 'name-fixes' into version/1.19.2 (#9862)

Fix some of the name files

Raycoms committed:
fix entity kill stat display and fix break event cancel (#9780)

fix entity kill stat display
and fix break event cancel

Raycoms committed:
add water clearing step, use spawn entity step, fix msg (#9790)

Properly clear water
Send the right msg on supply reset
Use the right entity spawn step

Raycoms committed:
Bunch o fixes (#9815)

Fix issue with builder
New banner pattern
Small quest fix
Fix weird forester loop

Raycoms committed:
Improve raider entity spawn position (#9802)

Improve raider entity spawn position
Fix dimensions of a few textures

Raycoms committed:
add new way to detect DO mining (#9825)

add new way to detect DO mining

Raycoms committed:
Immersive colony start (#9826)

Add data driven stories to abandoned supplycamps
Add data driven stories to abandoned supplyships
Add stories to abandoned colonies
Server-driven colony creation (pre-validation before townhall opening)
New UI for colony deletion/abandoning
New UI for colony reactivation
New UI for colony creation (includes entering name)
New translation strings with more immersive messages for all of these things
More relaxed placement of supplyship/camp taking the placeholders into account (will need some schematic adjustments possibly)
Fix issues with supplycamp/ship placement handling

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=9773) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=9773) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=9773) before we can accept your contribution.<br/>**5** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: Thodor12<br/>:white_check_mark: Nightenom<br/>:white_check_mark: someaddons<br/>:white_check_mark: MotionlessTrain<br/>:white_check_mark: Raycoms<br/>:x: imgbot[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ldtteam/minecolonies?pullRequest=9773) it.</sub>

Raycoms committed:
fixes

Raycoms committed:
hotfix 9828

Raycoms committed:
Pathfinding restructuring & Fixes (#9837)

Added new PathfindingUtils, which hold various common checks used in pathfinding
Refactored packages of pathfinding classes, they're no longer split
Removed/Cleaned up a lot of unused/old/legacy code
Extracted restriction zones to the specific path job types

There is now two different constructors for AbstractPathJob for two different types of pathfinding:
Start + range for exploring types which do now know their destination
Start + end for go to types which know where they end at
Cost computation for corner nodes is now based on the parent node, fixing stepup costs for stairs
Cost computation for ladders now is an entry-only cost, once on a ladder there is no extra to continue using it

Added a randomness cost to pathing options, giving each node a small random cost value additionally, which makes paths more natural and breaks off tie issues
Added the ability to explore additional nodes after finding a possible destination, to allow weak destination conditions to explore a bit further, those are then compared by node result score
Improved the block lookup cache, allowing it to keep previous values that are in bounds

Pathfinding debugging now displays heuristic and cost values
Pathfinding debug entity tracking now uses UUID isntead of storing players

Colony loaded chunks can now be accessed offthread
Fix heuristics/end node scores in serveral pathjobs
Fix stuckhandler sometimes resetting paths despite the citizens progressing
Fix pond search logic
Fix lumberjack tree search with restrictions failing under some conditions
Fix nonladder climbables not working

Raycoms committed:
Fix citizen references (#9844)

Fix citizen references used after removal
Fix citizens not unregistering

Raycoms committed:
fix at

Raycoms committed:
couple of fixes (#9845)

Fix citizens thinking they are at the hospital when they're not
Reduce miner pull-ahead to 4, (10 is a bit much)
Fix courier picking up stuff from builder (this will probably also improve builder pickup efficiency)
Fix exceptions when firing couriers with active tasks
Already add better DO placement handling

Raycoms committed:
Refactor stories system (#9841)

Move the stories from data pack to resource pack, because they're purely client side anyway and the previous design seemed potentially problematic for servers.
Parse story elements across multiple files, making it easy to add extra stories in packs.
Each file can have either a single element or a list of elements, of one or multiple types, so you can group them however seems to make sense. The content can either be a single entry or multiple entries that share the same biome filter.
Allow story elements to be filtered by biome (either by tag or exact).
Multiple filters are combined via OR. Specifying no filters means they're valid for all biomes.
Currently, only the abandoned colony names are actually filtered; some of the stories probably should be as well but I wanted some feedback/suggestions for particular groupings first.
Abandoned colonies use the biome where the town hall block sits, as before.
Loot supply camps/ships use a random biome (the idea being that the supplies may have been moved before you found them anyway).
Basic datagen for the files, mostly so there's auto-complete for the biome tags.

Raycoms committed:
Several fixes (#9857)

Move dyer research
Remove cobble from bonus blocks
Recruited citizens spawn at pos of visitor
Fix mourning
fix DO panel placement
Small pathing fix

Raycoms committed:
Fix tooltip appearing on settings where it shouldn't (#9848)

Fix issue when tooltips have an inactive reason, the tooltip would show when it shouldn't

Raycoms committed:
pathfinding heuristics (#9863)

Navigators now keep a long-time average of cost vs distance relation to allow adaptation of heuristics to the terrain (for reachability/nodes explored)

Pathjobs now re-evalute heuristic values used on reaching the target or when progress is going badly

Add a node limit hit callback to pathjobs to allow increasing a limit when conditions are met

Improved pathing stuck handler to not take action while an entity is progressing on a path

Improved pathing stuck handler to try walking in all 4 directions, instead of only opposite direction

Add new pathjob to move into a given direction for X range

Extend debug draw with more types of nodes, path reached nodes are now yellow instead of orange

Pathfinding costs are now applied in order, boni are multiplied first, mali are additive (reduced base for all by one due to this)

Navigator now keeps the pathresult of finished paths, to allow comparing new incomming ones

Fix block lookup cache sometimes shifting previous blockstates to wrong places

Raycoms committed:
compile fixes

Raycoms committed:
Several fixes (#9833)

Modified the hire trainee setting from a bool setting to a dedicated setting, this way the active state can be controlled
With this the hire trainee setting now becomes disabled when none of the different guard types are automatic hire
Modified some tags related to bamboo crafting recipes
Fixed a few other tags to make them more consistent
Added the option for the forester to strip bamboo blocks into stripped bamboo blocks

Raycoms committed:
fix compile issues

Raycoms committed:
fix sync of visited nodes

Raycoms committed:
dont upload to curse
